front_end/Scripts.js

```diff
@@ -1,32 +1,101 @@
+var mqtt = require('mqtt')
+var client  = mqtt.connect('mqtt://localhost')
+let $ = require('jquery')
+
+var stop = false;
+var frameCount = 0;
+var $results = $("#results");
+var fps, fpsInterval, startTime, now, then, elapsed;
+let x, y = false;  //this is so that the constoller doesn't send inputs when it's not in use 
+
 
 const background = document.querySelector('html')
 background.addEventListener('keydown', (e)=> {
-    console.log(e.key);
+    if(e.key == 'w'){
+        client.publish('/smartcar/control/throttle', '100')
+    }  
+    if(e.key == 's'){
+        client.publish('/smartcar/control/throttle', '-100')
+    }  
+    if (e.key == 'a'){
+        client.publish('/smartcar/control/steering', '-45')
+    }  
+    if (e.key == 'd'){
+        client.publish('/smartcar/control/steering', '45')
+    } 
+    console.log(e.key)
+})
+
+
+background.addEventListener('keyup', e => {
+    if(e.key == 'w'){
+        client.publish('/smartcar/control/throttle', '0')
+    }
+    if(e.key == 's'){
+        client.publish('/smartcar/control/throttle', '0') 
+    } 
+    if(e.key == 'a'){
+        client.publish('/smartcar/control/steering', '0') 
+    }  
+    if(e.key == 'd'){
+        client.publish('/smartcar/control/steering', '0')
+    }
 })
 
 
 
+function startAnimating(fps) {  //limit the output speed
+    fpsInterval = 1000 / fps;
+    then = Date.now();
+    startTime = then;
+    updateGamepad();
+}
+
+
 function updateGamepad(){
     requestAnimationFrame(updateGamepad)
-    let gp = navigator.getGamepads()[0];
-    if (gp !== null){
-        gp.buttons.forEach(button => {
-            if(button.pressed == true){
-                console.log(button, gp.buttons.indexOf(button));
+    now = Date.now();
+    elapsed = now - then;
+    if (elapsed > fpsInterval) {
+        then = now - (elapsed % fpsInterval);
+        let gp = navigator.getGamepads()[0];
+        if (gp !== null){
+            gp.buttons.forEach(button => {
+                if(button.pressed == true){
+                    console.log(button, gp.buttons.indexOf(button));
+                    client.publish('/smartcar/control/buttons', String(gp.buttons.indexOf(button)))
+                }
+            })
+            /*
+            if(gp.axes[0] > 0.15 || gp.axes[0] < -0.15) {
```

Comment:
That is pretty good, do we need left joystick to control anything for the car or do we take out of the code permanently? Or should we keep it if we are going to add something in the future.

Reply:
We only use the right joystick to control the car for now so that part is not needed for now. 
Should it be deleted?

Reply:
I guess we can delete it and if needed we can add it again later